### PR TITLE
node2vec: Add checks in order to avoid unexpected behavior and segfaults

### PR DESCRIPTION
**Pull request by fabratu**
This PR adds checks for node2vec (directed, has isolated nodes) in order to avoid segfaults (see #1011). Not included here, but on the roadmap: Adding the support of directed graphs for node2vec.

### CODE REVIEW

**Review by angriman — COMMENTED** [review]
I suggest to add some simple tests to check that errors are thrown in case of unsupported graphs (i.e., directed graphs, graphs with isolated nodes, and graphs with non-continuous node ids).

File: networkit/cpp/embedding/Node2Vec.cpp [hunk 1]
```diff
@@ -31,12 +31,21 @@ namespace NetworKit {
...
-        throw std::runtime_error("The node ids of the graph must be continuous.");
+    if (G.isDirected()) {
+        throw std::runtime_error("Current implementation can only deal with undirected graphs");
     }
 }
 
 void Node2Vec::run() {
+    if (G->numberOfNodes() != G->upperNodeIdBound()) {
+        throw std::runtime_error("The node ids of the graph must be continuous.");
+    }
```
[the remark this answers is not in the document]
Comment (fabratu) [reply networkit/cpp/embedding/Node2Vec.cpp:42]: My thought was: sanity checks for graph properties, which can be changed in place in a graph instance, don't necessarily need to be triggered in the constructor. In this way `run()` might succeed after the graph was changed and there is no need to create a new instance of the algortihm. For undirected/directed this is not suitable (therefore the check is placed in the constructor). 

But maybe it is advisable to not change the previous behavior in order not to break code.

[the remark this answers is not in the document]
Comment (angriman) [reply networkit/cpp/embedding/Node2Vec.cpp:42]: I think that it's ok to have these checks in the constructor and document which kind of input the algorithm expects. In `run()` we should assume that the input is fine. I would consider changing the graph between the constructor and `run()` a misuse of the algorithm and we should not be responsible to check that.

[the remark this answers is not in the document]
Comment (fabratu) [reply networkit/cpp/embedding/Node2Vec.cpp]: Thats also a way of looking at it. Probably the right call given how all other algorithms are implemented. Will change it.

File: networkit/cpp/embedding/Node2Vec.cpp [hunk 2]
```diff
@@ -31,9 +31,20 @@ namespace NetworKit {
...
         throw std::runtime_error("The node ids of the graph must be continuous.");
     }
+    bool hasIsolatedNodes = false;
+    G.parallelForNodes([&](const node u) {
+        if (G.degree(u) == 0) {
+            hasIsolatedNodes = true;
+        }
+    });
+    if (hasIsolatedNodes)
+        throw std::runtime_error("Isolated nodes are not allowed.");
```
Comment (angriman) [inline networkit/cpp/embedding/Node2Vec.cpp:47]: It's probably cleaner to use:
```suggestion
    for (node u : G.nodeRange())
        if (G.degree(u) == 0)
            std::runtime_error("Isolated nodes are not allowed.");
```

Comment (fabratu) [reply networkit/cpp/embedding/Node2Vec.cpp]: Exchanged `parallelForNodes` for `nodeRange`. However, testing for throws fails if the runtime_error is triggered inside a for-loop.

File: networkit/cpp/embedding/Node2Vec.cpp [hunk 3]
```diff
@@ -31,9 +31,18 @@ namespace NetworKit {
...
+    if (G.isDirected()) {
+        throw std::runtime_error("Current implementation can only deal with undirected graphs");
+    }
     if (G.numberOfNodes() != G.upperNodeIdBound()) {
         throw std::runtime_error("The node ids of the graph must be continuous.");
     }
+    bool hasIsolatedNodes = false;
+    for (node u : G.nodeRange())
+        if (G.degree(u) == 0)
+            hasIsolatedNodes = true;
```
Comment (angriman) [inline networkit/cpp/embedding/Node2Vec.cpp:43]: Weird that throwing inside the loop does not work. Anyway, I'd suggest to stop the loop once we find an isolated node:
```suggestion
        if (G.degree(u) == 0) {
            hasIsolatedNodes = true;
            break;
        }
```

**Review by angriman — APPROVED** [review]

**Pull request merged**